crates/red_knot_test/README.md

```diff
@@ -0,0 +1,430 @@
+
+Two kinds of assertions are supported, `# revealed:` (shown above) and `# error:`.
+
+### Assertion kinds
+
+#### revealed
+
+A `# revealed:` assertion should always be paired with a call to the `reveal_type` utility, which
+reveals (via a diagnostic) the inferred type of its argument (which can be any expression). The text
+after `# revealed:` must match exactly with the displayed form of the revealed type of that
+expression.
+
+The `reveal_type` function can be imported from the `typing` standard library module (or, for older
+Python versions, from the `typing_extensions` pseudo-standard-library module -- it is a third-party
+module, but typeshed, and thus type checkers also, treat it as part of the standard library):
+
+```py
+from typing import reveal_type
+
+reveal_type("foo")  # revealed: Literal["foo"]
+```
+
+For convenience, type checkers also pretend that `reveal_type` is a built-in, so that this import is
+not required. Using `reveal_type` without importing it issues a diagnostic warning that it was used
+without importing it, in addition to the diagnostic revealing the type of the expression.
+
+The `# revealed:` assertion must always match a revealed-type diagnostic, and will also match the
+undefined-reveal diagnostic, if present, so it's safe to use `reveal_type` in tests either with or
+without importing it. (Style preference is to usually not import it in tests, unless specifically
+testing something about the behavior of importing it.)
+
+#### error
+
+A comment beginning with `# error:` is an assertion that a type checker diagnostic will
+be emitted, with text span starting on that line. If the comment is simply `# error:`, this will
+match any diagnostic. The matching can be narrowed in three ways:
+
+- `# error: 8` requires that the matched diagnostic's text span begins on column 8 (one-indexed) of
+    this line.
+- `# error: [invalid-assignment]` requires that the matched diagnostic have the rule code
+    `invalid-assignment`.
+- `# error: "Some text"` requires that the matched diagnostic's full message contain the text `Some text`.
+
+Any combination of two or all three of these can be used, but they must go in order: first column,
+if present; then rule code, if present, then contains-text, if present. For example, an assertion
+using all three would look like `# error: 8 [invalid-assignment] "Some text"`.
+
+Error assertions in tests intended to test type checker semantics should use primarily rule-code
+assertions, with occasional contains-text assertions where needed to disambiguate.
+
+Some tests may exist in order to assert on the correct text-range or text of a diagnostic; these
+should use the appropriate assertions.
+
+### Assertion locations
+
+An assertion comment may be a line-trailing comment, in which case it applies to the line it is on:
+
+```py
+x: str = 1  # error: [invalid-assignment]
+```
+
+Or it may be a comment on its own line, in which case it applies to the next non-assertion-comment
+line:
```

Comment:
I find "non-assertion-comment" a bit confusing because at first I thought it meant a non-assertion _comment_, so maybe the next statement / code line?